### PR TITLE
Use ComplexHolder pointer in c10:Scalar

### PR DESCRIPTION
**Pull request by wenleix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53329 [Benchmark Only] Benchmark "almost trivial" move/copy constructor/assignment for Scalar
* #53318 [Benchmark Only] Remove move/copy constructor/assignment from Scalar
* **#53247 Use ComplexHolder pointer in c10:Scalar**
* #53246 Refactor ivalue::ComplexHolder as a top-level struct

Summary:

Test Plan:
USE_CUDA=0 BUILD_TEST=0 python setup.py install

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26810622](https://our.internmc.facebook.com/intern/diff/D26810622)

### CODE REVIEW

**Comment by wenleix** [comment]
Confirmed the Scalar size is reduced to 16 bytes with this PR. However, the following micro-benchmark shows ~1% **regression**. Will dig into. 

```python
from torch.utils.benchmark import Timer
counts = Timer(
    stmt="""auto out = t.add(a);""",
    setup="auto t = torch::ones({1}); auto a = torch::Scalar(3);",
    language="cpp",
).collect_callgrind(number=1_000)
print(counts)
```

Use ComplexHolder in Scalar (4bd9eebd66c9781d5ae2605954f614f2ff39234f) : 29,023,819
Baseline (79944f7ad9c6a30b9aed899d32a52513a54c17ce) : 28,706,820

File: c10/core/Scalar.h [hunk 1]
```diff
@@ -26,6 +27,49 @@ class C10_API Scalar {
  public:
   Scalar() : Scalar(int64_t(0)) {}
 
+  Scalar(const Scalar& rhs) : tag(rhs.tag), v(rhs.v) {
+      if (isComplex()) {
+        c10::raw::intrusive_ptr::incref(v.z);
+      }
+  }
+
+  Scalar(Scalar&& rhs) noexcept : tag(rhs.tag), v(rhs.v) {}
```
Comment (smessmer) [inline c10/core/Scalar.h:36]: this is likely a bug causing a double free. After this, you now have two Scalar objects that both are going to destroy the same complex. Could be fixed by setting the *this scalar to not store a complex anymore similar to what you're doing in the move assignment operator

**Review by smessmer — APPROVED** [review]

File: c10/core/Scalar.h (hunk 1 above)
Comment (smessmer) [reply c10/core/Scalar.h:36]: as a side note: Having all four copy/move constructors and assignment operators implemented explicitly is some code duplication with the chance of bugs like this one. There's a nice pattern where you can implement all four of those in terms of just one implementation of `swap`, and then you have all the refcounting or destroying of state in just one function.

File: c10/core/Scalar.h [hunk 2]
```diff
@@ -26,6 +27,49 @@ class C10_API Scalar {
...
+    tag = rhs.tag;
+    v = rhs.v;
+
+    if (isComplex()) {
+        c10::raw::intrusive_ptr::incref(v.z);
+    }
+    return *this;
+  }
+
+  C10_ALWAYS_INLINE Scalar& operator=(Scalar&& rhs) & noexcept {
```
Comment (smessmer) [inline c10/core/Scalar.h:55]: any specific reason for using `C10_ALWAYS_INLINE` here? In almost all cases, the compiler knows better and forcing them to inline is a bad idea. Without this, for example, the compiler can make different decisions like say inlining it in server code but not on mobile.

File: c10/core/Scalar.h (hunk 1 above)
Comment (wenleix) [reply c10/core/Scalar.h:36]: Nice catch! The RHS complex will be cleaned up in move assignment but forget to to the same cleanup in move constructor.

Comment (wenleix) [reply c10/core/Scalar.h:36]: @smessmer : Nice catch! The RHS complex will be cleaned up in move assignment but forget to to the same cleanup in move constructor.

File: c10/core/Scalar.h (hunk 2 above)
Comment (wenleix) [reply c10/core/Scalar.h:55]: @smessmer : Got it. Originally when I start to write these move/copy methods, I followed the corresponding implementation in `IValue` and didn't reove `C10_ALWAYS_INLINE
` https://github.com/pytorch/pytorch/blob/56f8379802960b6cbd9973a02cb03efba14dfb7e/aten/src/ATen/core/ivalue.h#L197

**Comment by wenleix** [comment]
Superseded by the codemod based solution (https://github.com/pytorch/pytorch/pull/53583) due to overhead of non-trivial constructors.

**Pull request closed without merging**